### PR TITLE
CMake set

### PR DESCRIPTION
**Pull request by 55-AA**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -112,9 +112,7 @@ if(PYTHON_BINDINGS)
       set(PYTHON_LIBRARIES "$ENV{PYTHON_LIBRARIES}")
     endif()
     if(NOT PYTHON_INCLUDE_DIRS AND NOT PYTHON_LIBRARIES)
-        find_package(PythonLibs 2.7 REQUIRED)
-      else()
-        message(FATAL_ERROR "Inconsistent PYTHON_INCLUDE_DIRS and PYTHON_LIBRARIES")
```
Comment (JonathanSalwan) [inline CMakeLists.txt]: Don't erase these lines.

File: src/libtriton/CMakeLists.txt [hunk 2]
```diff
@@ -8,17 +8,16 @@ endif()
 # Default flags
 # To fix old python version bug http://bugs.python.org/issue21958
 if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DHAVE_ROUND /MT")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DHAVE_ROUND -D_CRT_SECURE_NO_WARNINGS /MT")
 endif()
-
 if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
```
Comment (JonathanSalwan) [inline src/libtriton/CMakeLists.txt]: Don't remove this empty line.

File: src/libtriton/api/api.cpp [hunk 3]
```diff
@@ -367,33 +366,38 @@ namespace triton {
...
-      throw triton::exceptions::API("API::initEngines(): No enough memory.");
+      this->z3Interface = new(std::nothrow) triton::ast::Z3Interface(this->symbolic);
+      if (this->z3Interface == nullptr)
+        break;
 
-    this->z3Interface = new(std::nothrow) triton::ast::Z3Interface(this->symbolic);
-    if (this->z3Interface == nullptr)
-      throw triton::exceptions::API("API::initEngines(): No enough memory.");
+      return;
+    }while(0);
```
Comment (JonathanSalwan) [inline src/libtriton/api/api.cpp]: I don't really like this change.

File: src/libtriton/ast/ast.cpp [hunk 4]
```diff
@@ -3122,435 +3122,387 @@ namespace triton {
 namespace triton {
   namespace ast {
 
+    void checkNode(AbstractNode* node){
```
Comment (JonathanSalwan) [inline src/libtriton/ast/ast.cpp]: `static inline void`

File: src/libtriton/ast/ast.cpp [hunk 5]
```diff
@@ -3122,435 +3122,387 @@ namespace triton {
...
+    if (node == nullptr)
```
Comment (JonathanSalwan) [inline src/libtriton/ast/ast.cpp]: `if` alignment.

File: CMakeLists.txt (hunk 1 above)
Comment (55-AA) [inline CMakeLists.txt]: Ok.But if so, python path cannot be set from cmake args.

Comment (JonathanSalwan) [inline CMakeLists.txt]: Why?

File: src/libtriton/api/api.cpp (hunk 3 above)
Comment (55-AA) [inline src/libtriton/api/api.cpp]: In VS2015, use MinSize build option, the assembly code of the initEngines() was less 247 bytes than old. The old code were 757 bytes, the new were 510 bytes, there was 32% saved. In linux src, many MACRO were do-while style. If you don't really like it, I'll revert it.

File: src/libtriton/ast/ast.cpp (hunk 4 above)
Comment (55-AA) [inline src/libtriton/ast/ast.cpp]: Sure, static should be in. But  may I don't use inline. It shall cause too many repeat code in binary.

File: src/libtriton/ast/ast.cpp (hunk 5 above)
Comment (55-AA) [inline src/libtriton/ast/ast.cpp]: Really.

File: CMakeLists.txt (hunk 1 above)
Comment (55-AA) [inline CMakeLists.txt]: D:\WorkData\Instrumentation\Triton\build>cmake -G "Visual Studio 14 2015 Win64" -DPYTHON_INCLUDE_DIRS="C:/Pyth
de" -DPYTHON_LIBRARIES="C:/Python27/libs/python27.lib" -DBOOST_ROOT="D:/OpenSrc/boost_1_63_0" -DZ3_INCLUDE_DIR
Src/z3-4.5.0-x64-win/include" -DZ3_LIBRARIES="D:/OpenSrc/z3-4.5.0-x64-win/bin/libz3.lib" -DCAPSTONE_INCLUDE_DI
nSrc/capstone-3.0.5-rc2-win64/include" -DCAPSTONE_LIBRARIES="D:/OpenSrc/capstone-3.0.5-rc2-win64/capstone.lib"
CMake Error at CMakeLists.txt:117 (message):
  Inconsistent PYTHON_INCLUDE_DIRS and PYTHON_LIBRARIES


-- Configuring incomplete, errors occurred!
See also "D:/WorkData/Instrumentation/Triton/build/CMakeFiles/CMakeOutput.log".

D:\WorkData\Instrumentation\Triton\build>

Comment (55-AA) [inline CMakeLists.txt]: D:\WorkData\Instrumentation\Triton\build>cmake -G "Visual Studio 14 2015 Win64" -DPYTHON_INCLUDE_DIRS="C:/Python27/include" -DPYTHON_LIBRARIES="C:/Python27/libs/python27.lib" -DBOOST_ROOT="D:/OpenSrc/boost_1_63_0" -DZ3_INCLUDE_DIRS="D:/OpenSrc/z3-4.5.0-x64-win/include" -DZ3_LIBRARIES="D:/OpenSrc/z3-4.5.0-x64-win/bin/libz3.lib" -DCAPSTONE_INCLUDE_DIRS="D:/OpenSrc/capstone-3.0.5-rc2-win64/include" -DCAPSTONE_LIBRARIES="D:/OpenSrc/capstone-3.0.5-rc2-win64/capstone.lib" ..
CMake Error at CMakeLists.txt:117 (message):
  Inconsistent PYTHON_INCLUDE_DIRS and PYTHON_LIBRARIES


-- Configuring incomplete, errors occurred!
See also "D:/WorkData/Instrumentation/Triton/build/CMakeFiles/CMakeOutput.log".

D:\WorkData\Instrumentation\Triton\build>

Comment (55-AA) [inline CMakeLists.txt]: If usr set PYTHON params, the line 117 is always true.  It shall break building.

Comment (55-AA) [inline CMakeLists.txt]: #505

**Comment by JonathanSalwan** [comment]
We will accept your `cpp` part after merging #507. About the `static void`, we prefer to define it as `static inline void` for the reason that billion of AST nodes are created per minute and we prefer to improve the speed at the runtime instead of the size of the libtriton.

**Comment by JonathanSalwan** [comment]
Okay, so, now that all patchs about cmake are merged into `master`, can you update this PR to only keep modification about `api.cpp` and `ast.cpp` ?

**Comment by 55-AA** [comment]
No update.

**Comment by 1n598** [comment]
Guys just to let you know...sometimes it's good to check manually the Python command that is executed. For me the issue was that the Path was not set properly for python, so the variable for the destination was always empty.

**Comment by JonathanSalwan** [comment]
go ahead for a pr then

**Comment by 1n598** [comment]
Once I manage to fully compile it I will. So far I have other issues compiling this on a Windows 7 32 bits. Will keep everyone posted.

**Pull request closed without merging**